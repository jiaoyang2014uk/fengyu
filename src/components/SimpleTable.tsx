import {defineComponent} from "vue";
import TableHeader from "./TableHeader";
import TableBody from "./TableBody";
import Pagination from "./Pagination";
import {useColumnsMethod, useFormattedDataMethod, usePaginationMethod} from '../common/hooks'
import {info, trace} from "../common/utils";
import '../common/Table.less';
import type {TableProps, SortOptions} from "../common/types";
import {tableProps} from "../common/const";

export default defineComponent({
    name: "SimpleTable",
    props: tableProps,
    setup(props: TableProps, {attrs, emit, slots}) {
        const {headerColumns, bodyColumns} = useColumnsMethod(props.columns, slots);
        const formattedData = useFormattedDataMethod(props);
        const total = usePaginationMethod(props.paginationOptions, props.data);
        const onSortChange = (sortOptions: SortOptions) => {
            trace(
                'SimpleTable: ',
                `onSortChange emit updateSortOptions with sortOptions: ${JSON.stringify(sortOptions)}`
            )
            emit('updateSortOptions', sortOptions);
        };
        const onUpdatePage = (page: number) => {
            const options = {
                ...props.paginationOptions,
                page
            };
            trace(
                'SimpleTable: ',
                `onUpdatePage emit updatePaginationOptions with options: ${JSON.stringify(options)}`
            )
            emit('updatePaginationOptions', options);
        };
        return () => {
            info(
                'SimpleTable: ',
                `columns: ${JSON.stringify(props.columns)}, data: ${JSON.stringify(props.data)}, sortOptions: ${JSON.stringify(props.sortOptions)}, paginationOptions: ${JSON.stringify(props.paginationOptions)}`
            )
            return (
                <div class='table'>
                    <table class="table__table">
                        <TableHeader columns={headerColumns} sortOptions={props.sortOptions}
                                     onUpdateSortOptions={onSortChange}/>
                        <TableBody columns={bodyColumns} data={formattedData.value}/>
                    </table>
                    <Pagination v-show={props.paginationOptions.enable}
                                limit={props.paginationOptions.limit}
                                page={props.paginationOptions.page}
                                total={total.value}
                                onUpdatePage={onUpdatePage}/>
                </div>
            );
        };
    },
});
