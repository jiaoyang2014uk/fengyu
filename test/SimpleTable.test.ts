import {mount} from "@vue/test-utils";
import {expect, test} from "vitest";
import SimpleTable from "../src/components/SimpleTable";

const mountTable = (options?: Record<string, unknown>) => mount(SimpleTable, options);

test("mount component without props", () => {
    const wrapper = mountTable();

    expect(wrapper.html()).toMatchSnapshot();
});

test('props with columns and data', async () => {
    const wrapper = mountTable({
        props: {
            columns: [
                {header: 'Name', dataIndex: 'name'},
                {header: 'Age', dataIndex: 'age', sortable: true}
            ],
            data: [
                {name: 'Bob', age: '27'},
                {name: 'Tom', age: '20'},
                {name: 'Alice', age: '23'}
            ]
        }
    });

    expect(wrapper.html()).toMatchSnapshot();
});

test('header slot and body slot', async () => {
    const wrapper = mountTable({
        props: {
            columns: [
                {header: 'Name', dataIndex: 'name'},
                {header: 'Age', dataIndex: 'age', sortable: true}
            ],
            data: [
                {name: 'Bob', age: '27'},
                {name: 'Tom', age: '20'},
                {name: 'Alice', age: '23'}
            ]
        },
        slots: {
            header__name: 'NAME',
            body__age: '<template v-slot:body__age="obj">\n' +
                '            {{ obj.data.name + \'(\' + obj.data.age + \')\' }}\n' +
                '        </template>'
        },
    });

    expect(wrapper.html()).toMatchSnapshot();
});

