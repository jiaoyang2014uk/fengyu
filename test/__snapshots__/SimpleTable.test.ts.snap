// Vitest Snapshot v1

exports[`header slot and body slot 1`] = `
"<div class=\\"table\\">
  <table class=\\"table__table\\">
    <thead>
      <tr>
        <th class=\\"table__header-cell\\"><span class=\\"table__header-cell__value\\">NAME</span><span class=\\"table__header-cell_sort\\" style=\\"display: none;\\"></span></th>
        <th class=\\"table__header-cell\\"><span class=\\"table__header-cell__value\\">Age</span><span class=\\"table__header-cell_sort\\"></span></th>
      </tr>
    </thead>
    <tbody>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Bob</span></td>
        <td class=\\"table__body-cell\\"><span>Bob(27)</span></td>
      </tr>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Tom</span></td>
        <td class=\\"table__body-cell\\"><span>Tom(20)</span></td>
      </tr>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Alice</span></td>
        <td class=\\"table__body-cell\\"><span>Alice(23)</span></td>
      </tr>
    </tbody>
  </table>
  <div class=\\"pagination\\" style=\\"display: none;\\"><span class=\\"pagination__total\\">Total 3 items</span><span class=\\"pagination__previous disabled\\">Previous</span><span class=\\"pagination__next disabled\\">Next</span><span class=\\"pagination__page\\">( 1 / 1 )</span><span class=\\"pagination__goto\\">Go to<input type=\\"text\\" class=\\"pagination__goto-input\\"></span></div>
</div>"
`;

exports[`mount component 1`] = `
"<table>
  <thead>
    <th>
    <td></td>
    </th>
  </thead>
  <tbody>
    <tr>
      <td>hello table</td>
    </tr>
  </tbody>
</table>"
`;

exports[`mount component without props 1`] = `
"<div class=\\"table\\">
  <table class=\\"table__table\\">
    <thead>
      <tr></tr>
    </thead>
    <tbody></tbody>
  </table>
  <div class=\\"pagination\\" style=\\"display: none;\\"><span class=\\"pagination__total\\">Total 0 items</span><span class=\\"pagination__previous disabled\\">Previous</span><span class=\\"pagination__next disabled\\">Next</span><span class=\\"pagination__page\\">( 1 / 0 )</span><span class=\\"pagination__goto\\">Go to<input type=\\"text\\" class=\\"pagination__goto-input\\"></span></div>
</div>"
`;

exports[`props with columns and data 1`] = `
"<div class=\\"table\\">
  <table class=\\"table__table\\">
    <thead>
      <tr>
        <th class=\\"table__header-cell\\"><span class=\\"table__header-cell__value\\">Name</span><span class=\\"table__header-cell_sort\\" style=\\"display: none;\\"></span></th>
        <th class=\\"table__header-cell\\"><span class=\\"table__header-cell__value\\">Age</span><span class=\\"table__header-cell_sort\\"></span></th>
      </tr>
    </thead>
    <tbody>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Bob</span></td>
        <td class=\\"table__body-cell\\"><span>27</span></td>
      </tr>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Tom</span></td>
        <td class=\\"table__body-cell\\"><span>20</span></td>
      </tr>
      <tr class=\\"table__body-row\\">
        <td class=\\"table__body-cell\\"><span>Alice</span></td>
        <td class=\\"table__body-cell\\"><span>23</span></td>
      </tr>
    </tbody>
  </table>
  <div class=\\"pagination\\" style=\\"display: none;\\"><span class=\\"pagination__total\\">Total 3 items</span><span class=\\"pagination__previous disabled\\">Previous</span><span class=\\"pagination__next disabled\\">Next</span><span class=\\"pagination__page\\">( 1 / 1 )</span><span class=\\"pagination__goto\\">Go to<input type=\\"text\\" class=\\"pagination__goto-input\\"></span></div>
</div>"
`;
